### ISSUE TITLE
[Parity 15/16] Optimizers and LR schedulers completeness - LBFGS/SparseAdam/RAdam + full scheduler suite

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#224]
Part of epic #209.

## Summary

`FusedOptimizer.cs` exists. The full PyTorch optimizer + learning-rate-scheduler surface is missing at the tensor-library layer. This covers: Adam/AdamW/SGD-momentum (likely present), LBFGS, SparseAdam, RAdam, NAdam, Adamax, ASGD, Rprop, and the complete LR-scheduler family.

Upstream AiDotNet core has its own optimizer module. This issue makes sure the *tensor-layer* fused kernels exist for every optimizer that benefits from one.

### Optimizers
- `SGD` (with momentum, dampening, weight_decay, nesterov, maximize)
- `Adam`, `AdamW`, `NAdam`, `RAdam`, `Adamax`
- `Adadelta`, `Adagrad`
- `RMSprop`
- `LBFGS`
- `SparseAdam` (Adam variant for sparse gradients)
- `ASGD` (Averaged SGD)
- `Rprop` (Resilient Backprop — small-batch classical)
- Newer experimental: `Lion`, `Shampoo`, `D-Adaptation`, `Prodigy` (via pytorch-optimizer)

### LR schedulers (`torch.optim.lr_scheduler`)
- `StepLR`, `MultiStepLR`
- `ExponentialLR`
- `CosineAnnealingLR`, `CosineAnnealingWarmRestarts`
- `CyclicLR` (triangular, triangular2, exp_range)
- `OneCycleLR` (linear + cosine anneal variants)
- `ReduceLROnPlateau`
- `LambdaLR`, `MultiplicativeLR`
- `ConstantLR`, `LinearLR`
- `ChainedScheduler`, `SequentialLR`
- `PolynomialLR`
- Warmup variants

### Utilities
- `lr_scheduler.get_last_lr()`, `.step()` semantics
- Per-param-group config
- State dict save/load

## Current state in the repo

- `src/AiDotNet.Tensors/Helpers/Optimized/FusedOptimizer.cs` — fused kernel present. Coverage unclear.
- Upstream AiDotNet core has optimizer classes. This issue complements them by ensuring every optimizer has a fused kernel path here.

## How we beat PyTorch

1. **Fused kernels for every optimizer** that benefits: AdamW, SGD-momentum, LAMB, Lion, RAdam, NAdam, Adamax — one kernel launch per parameter group instead of six. PyTorch ships `foreach_` and `fused_` variants for some; we ship them for all, auto-selected.
2. **Mixed-precision optimizer state**: BF16 / FP8 optimizer moments with FP32 master params, already half of our `MixedPrecisionTrainingLoop`. Formalize the pattern across every optimizer. Target 2–4× memory reduction on large models. PyTorch's `torch.optim._multi_tensor` is partial.
3. **LBFGS with our linalg**: line-search uses our `solve_triangular`; hessian update uses our banded-matrix paths. PyTorch's LBFGS is slow; we target 2× by leveraging BLAS for the two-loop recursion.
4. **SparseAdam on sub-byte sparsity**: when gradients come out of a 2:4 structured path (sparse issue), SparseAdam update touches only the live channels in the packed representation. Unique.
5. **Scheduler DSL**: chain schedulers via `SequentialLr`, `ChainedLr`, and also a new `ScheduleBuilder` fluent API for common recipes: `warmup(1000) → cosine(90k) → linear_decay(5k)`. PyTorch requires multiple constructor calls.
6. **State-dict save/load compatible with PyTorch**: users can load PyTorch checkpoints including optimizer state. Combined with the serialization issue, this gives us full `.pt` interop.
7. **ZeRO-integrated optimizers**: optimizer state sharding flows through the distributed issue automatically. Any optimizer here works with FSDP/ZeRO without special support.
8. **CUDA Graph capture of optimizer step**: already hinted by our `CudaGraphScope`; make sure every optimizer is capture-safe. PyTorch's `torch.cuda.make_graphed_callables` requires manual stepping through traps.

### Optimizer fused-kernel coverage (CPU AVX-512 + CUDA)
- [ ] SGD (momentum, dampening, wd, nesterov)
- [ ] Adam / AdamW (with amsgrad, maximize)
- [ ] NAdam, RAdam, Adamax
- [ ] Adadelta, Adagrad
- [ ] RMSprop (with centered variant)
- [ ] LBFGS (two-loop recursion using BLAS)
- [ ] SparseAdam
- [ ] ASGD
- [ ] Rprop
- [ ] Lion, LAMB (modern / large-batch)
- [ ] Shampoo / D-Adaptation / Prodigy (bonus)

### Mixed-precision optimizer state
- [ ] BF16 moments / FP32 master params across all
- [ ] FP8 moments where the optimizer is numerically forgiving (Lion, Adafactor)

### LR schedulers
- [ ] `StepLr`, `MultiStepLr`
- [ ] `ExponentialLr`, `PolynomialLr`
- [ ] `CosineAnnealingLr`, `CosineAnnealingWarmRestarts`
- [ ] `CyclicLr` (triangular, triangular2, exp_range)
- [ ] `OneCycleLr`
- [ ] `ReduceLrOnPlateau`
- [ ] `LambdaLr`, `MultiplicativeLr`
- [ ] `ConstantLr`, `LinearLr`
- [ ] `ChainedScheduler`, `SequentialLr`
- [ ] `ScheduleBuilder` fluent DSL (bonus)

### Plumbing
- [ ] Per-param-group config
- [ ] State-dict save/load (PyTorch-compatible via serialization issue)
- [ ] CUDA Graph–safe `Step()` for every optimizer
- [ ] ZeRO/FSDP integration hooks

## Acceptance criteria

- Every optimizer passes a convergence test on a standard mini-benchmark (linear regression, MLP on MNIST — tiny but sufficient).
- Fused kernel produces bit-identical results to the non-fused reference up to FP non-associativity.
- Mixed-precision variant stays within 2× the loss of FP32 baseline on a standard model training run.
- LBFGS benchmark ≥ 2× PyTorch LBFGS on a Rosenbrock + a logistic-regression problem.
- All schedulers: step-by-step test vs PyTorch reference.
- PyTorch state-dict load: read a `.pt` optimizer state produced by PyTorch Adam, step once, compare trajectory.
- CUDA Graph capture: `make_graphed_training_step` test green.

## Implementation notes

- `src/AiDotNet.Tensors/Engines/Optimization/Optimizers/` for fused kernels.
- `src/AiDotNet.Tensors/Engines/Optimization/Schedulers/` for schedulers.
- Reuse `Helpers/Optimized/FusedOptimizer.cs` as the base for fused variants.
- GPU bindings via existing `Engines/DirectGpu/*/` pattern.
- Each fused kernel registered with `AutotuneCache` for per-shape tuning.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

`Helpers/Optimized/FusedOptimizer.cs` — fused kernel present, coverage unclear. Upstream AiDotNet core has optimizer classes.

## 2. Scope — full inventory

### Optimizers (fused CPU AVX-512 + CUDA kernel coverage)
```
SGD (momentum, dampening, weight_decay, nesterov, maximize)
Adam / AdamW (amsgrad, maximize)
NAdam, RAdam, Adamax
Adadelta, Adagrad
RMSprop (centered variant)
LBFGS (two-loop recursion using BLAS)
SparseAdam
ASGD, Rprop
Lion, LAMB
Shampoo / D-Adaptation / Prodigy (bonus)
```

### LR schedulers
```
StepLr, MultiStepLr
ExponentialLr, PolynomialLr
CosineAnnealingLr, CosineAnnealingWarmRestarts
CyclicLr (triangular, triangular2, exp_range)
OneCycleLr
ReduceLrOnPlateau
LambdaLr, MultiplicativeLr
ConstantLr, LinearLr
ChainedScheduler, SequentialLr
ScheduleBuilder fluent DSL (bonus)
```

### Plumbing
```
Per-param-group config
State-dict save/load (PyTorch-compatible via #218)
CUDA Graph-safe Step() for every optimizer
ZeRO/FSDP integration hooks (#215)
```

**Total: ~30 ops.**

## 3. Win vectors

1. **Fused kernels for every optimizer** — one kernel launch per param group. PyTorch ships `fused_` for some; we ship for all.
2. **Mixed-precision optimizer state**: BF16/FP8 moments + FP32 master params. Target 2–4× memory reduction.
3. **LBFGS using BLAS** — two-loop recursion leverages our GEMV paths. Target 2× PyTorch LBFGS.
4. **SparseAdam on 2:4 structured sparsity** (#221) — update touches only live channels in packed repr. Unique.
5. **Scheduler DSL**: `ScheduleBuilder.Warmup(1000).Cosine(90k).LinearDecay(5k)` fluent.
6. **State-dict PyTorch-compatible** via #218 — load `.pt` optimizer state.
7. **ZeRO-integrated optimizers** auto via #215.
8. **CUDA Graph capture of optimizer step** — every optimizer capture-safe.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| AdamW fused step | ≥ parity PyTorch `fused_`; ≥1.5× PyTorch non-fused |
| LBFGS on Rosenbrock + logistic regression | ≥2× PyTorch LBFGS |
| Mixed-precision BF16 moments | 2× memory reduction, ≤2× loss over fp32 baseline |
| All schedulers step-by-step | parity PyTorch reference |
| `.pt` optimizer state load | round-trip parity |
| CUDA Graph capture | works for every optimizer; ≥20% training-loop speedup |
| All 6 GPU backends | CPU + CUDA required; others best-effort |

## 5. File inventory

**New (~55):**
- `src/AiDotNet.Tensors/Engines/Optimization/Optimizers/` — one file per optimizer (~18)
- `src/AiDotNet.Tensors/Engines/Optimization/Schedulers/` — one file per scheduler (~14)
- Fused kernels per backend (~20)
- `ScheduleBuilder.cs` DSL

**Modified (~6):** `FusedOptimizer.cs` base, `OpRegistry.cs`, `AutotuneCache.cs` (per-shape tuning), `LayerPrecisionPolicy.cs` (mixed-precision hooks).

**Tests (~25):** convergence per optimizer on linear reg + MLP-MNIST, bit-equality fused vs non-fused (up to FP non-assoc), mixed-precision within 2× loss, per-scheduler step parity, `.pt` state load, CUDA Graph capture.

**Benchmarks (~5):** `OptimizerStepBenchmarks`, `LBFGSBenchmarks`, `MixedPrecisionOptimBenchmarks`, `SchedulerBenchmarks`, `CudaGraphOptimizerBenchmarks`.

## 6. Acceptance gate

1. Every optimizer passes convergence test.
2. Fused produces bit-identical results to non-fused (up to FP non-associativity).
3. Mixed-precision variant ≤ 2× fp32 loss.
4. LBFGS ≥ 2× PyTorch.
5. All schedulers parity PyTorch step-by-step.
6. `.pt` state load round-trips.
7. CUDA Graph capture works for every optimizer.

## 7. Timeline — ~4 weeks

W1 SGD/Adam/AdamW/NAdam/RAdam/Adamax fused kernels.
W2 LBFGS + SparseAdam + ASGD/Rprop + Adadelta/Adagrad/RMSprop.
W3 Lion/LAMB + Shampoo/Prodigy + all schedulers.
W4 Mixed-precision + CUDA Graph capture + PR.

## 8. Dependencies

- **Blocks**: end-to-end training demos.
- **Blocked by**: #218 (PyTorch state-dict load for compatibility), #215 (ZeRO integration — non-blocking).

## 9. Risks

- **LBFGS two-loop recursion** with line search can misbehave on non-smooth problems — follow PyTorch's inf_f guard.
- **CUDA Graph capture + allocator** — deterministic allocator state required; co-develop with #225.
- **Shampoo/D-Adaptation/Prodigy** are research-grade — ship as experimental.

---

*Part of epic #209. Plan ratified 2026-04-19.*
